dace/sdfg.py

```diff
@@ -530,23 +528,49 @@ def constants(self):
         if self._parent_sdfg is not None:
             result.update(self._parent_sdfg.constants)
 
-        #result.update(self._constants)
-        result.update(self.constants_prop)
+        def cast(dtype: dt.Data, value: Any):
+            """ Cast a value to the given data type. """
+            if isinstance(dtype, dt.Array):
+                return value
+            elif isinstance(dtype, dt.Scalar):
+                return dtype.dtype(value)
+            raise TypeError('Unsupported data type %s' % dtype)
+
+        result.update({k: cast(*v) for k, v in self.constants_prop.items()})
         return result
 
-    def add_constants(self, new_constants: Dict[str, Any]):
+    def add_constant(self, name: str, value: Any):
```

Comment:
I guess we cannot enforce that users don't directly update the constants dictionary, but we hope that they do this by convention?

Reply:
We can hide away `add_constants` and make it an internal function, or just not exist at all. Your choice

Reply:
Why don't we let `add_constants` call `add_constant`?

Reply:
Sure. I can quickly write this up and let you know

Reply:
Fixed now